### PR TITLE
Master mrp bom structure and cost imp hch

### PR DESCRIPTION
**Pull request by hch-odoo**
task : https://www.odoo.com/web?#id=1857701&action=327&model=project.task&view_type=form&menu_id=4720

pad : https://pad.odoo.com/p/r.afcc15fc299bf803c3b466a7c44a6662

### CODE REVIEW

File: addons/mrp/report/mrp_report_bom_structure.py [hunk 1]
```diff
@@ -98,18 +99,33 @@ def _get_bom(self, bom_id=False, product_id=False, line_qty=False, line_id=False
...
+    def _has_attachments(self, bom):
+        has_attachments = False
+        for line in bom.bom_line_ids:
+            if line.has_attachments:
+                has_attachments = True
+                break;
+            if line.child_bom_id:
+                has_attachments = self._has_attachments(line.child_bom_id)
+                if has_attachments:
+                    break;
```
Comment (sle-odoo) [inline addons/mrp/report/mrp_report_bom_structure.py]: maybe make the some logic form bom line and sub bom lines
maybe this should be an helper on the mrp bom model?

File: addons/mrp/report/mrp_report_bom_structure.py [hunk 2]
```diff
@@ -98,18 +99,33 @@ def _get_bom(self, bom_id=False, product_id=False, line_qty=False, line_id=False
...
             'currency': self.env.user.company_id.currency_id,
             'product': product,
+            'code': bom.code,
             'price': product.uom_id._compute_price(product.standard_price, bom.product_uom_id) * bom_quantity,
             'total': sum([op['total'] for op in operations]),
             'level': level or 0,
             'operations': operations,
             'operations_cost': sum([op['total'] for op in operations]),
+            'attachments': self.env['mrp.document'].search(['|',
+            '&', ('res_model', '=', 'product.product'), ('res_id', '=', product.id), '&', ('res_model', '=', 'product.template'), ('res_id', '=', product.product_tmpl_id.id)]),
```
Comment (sle-odoo) [inline addons/mrp/report/mrp_report_bom_structure.py]: wrong indent

File: addons/mrp/report/mrp_report_bom_structure.xml [hunk 3]
```diff
@@ -24,17 +24,29 @@
...
                                     </tr>
                                 </thead>
                                 <tbody>
                                     <tr>
                                         <td>
                                             <span><a t-if="data['report_type'] == 'html'" t-att-data-res-id="data['product'].id" t-att-data-model="'product.product'" class="o_mrp_bom_action"><t t-esc="data['bom_prod_name']"/></a><t t-else="" t-esc="data['bom_prod_name']"/></span>
                                         </td>
+                                        <td>
+                                          <t t-if="data['report_type'] == 'html'">
+                                            <span>
```
Comment (sle-odoo) [inline addons/mrp/report/mrp_report_bom_structure.xml]: why indenting with two spaces here?

File: addons/mrp/report/mrp_report_bom_structure.py (hunk 2 above)
Comment (sle-odoo) [reply addons/mrp/report/mrp_report_bom_structure.py]: you're changing this line at rev 6c8654fbd9d2aa8d430f20bc03db39200c7e110f

please fix these commits

File: addons/mrp/report/mrp_report_bom_structure.py [hunk 4]
```diff
@@ -191,6 +191,7 @@ def get_sub_lines(bom, product_id, line_qty, line_id, level):
                     'prod_cost': bom_line['prod_cost'],
                     'bom_cost': bom_line['total'],
                     'level': bom_line['level'],
+                    'child_bom': bom_line['child_bom'],
```
Comment (sle-odoo) [inline addons/mrp/report/mrp_report_bom_structure.py]: why this change in this commit?

File: addons/mrp/static/src/js/mrp_bom_report.js [hunk 5]
```diff
@@ -109,11 +109,10 @@ var MrpBomReport = stock_report_generic.extend({
             return $(el).data('id');
         });
         framework.blockUI();
-        var reportname = 'mrp.report_bom_structure?docids=' + this.given_context.active_id;
+        var reportname = 'mrp.report_bom_structure?docids=' + this.given_context.active_id + '&report_type=' + this.given_context.report_type;
         if (! $(ev.currentTarget).hasClass('o_mrp_bom_print_unfolded')) {
             reportname += '&quantity=' + (this.given_context.searchQty || 1) +
-                          '&childs=' + JSON.stringify(childBomIDs) +
-                          '&report_type=' + this.given_context.report_type;
+                          '&childs=' + JSON.stringify(childBomIDs);
```
Comment (sle-odoo) [inline addons/mrp/static/src/js/mrp_bom_report.js]: i don't get this change, report_type was correctly appended at line 116?

File: addons/mrp/report/mrp_report_bom_structure.py [hunk 6]
```diff
@@ -142,6 +155,7 @@ def _get_bom_lines(self, bom, bom_quantity, product, line_id, level):
                 'phantom_bom': line.child_bom_id and line.child_bom_id.type == 'phantom' or False,
                 'attachments': self.env['mrp.document'].search(['|', '&',
                     ('res_model', '=', 'product.product'), ('res_id', '=', line.product_id.id), '&', ('res_model', '=', 'product.template'), ('res_id', '=', line.product_id.product_tmpl_id.id)]),
+
```
Comment (sle-odoo) [inline addons/mrp/report/mrp_report_bom_structure.py]: useless diff

File: addons/mrp/report/mrp_report_bom_structure.py [hunk 7]
```diff
@@ -38,8 +38,9 @@ def get_html(self, bom_id=False, searchQty=1, searchVariant=False):
         return res
 
     @api.model
-    def get_bom(self, bom_id=False, product_id=False, line_qty=False, line_id=False, level=False):
+    def get_bom(self, bom_id=False, product_id=False, line_qty=False, line_id=False, level=False, has_attachments=False):
```
Comment (amoyaux) [inline addons/mrp/report/mrp_report_bom_structure.py]: What is the purpose of this kwargs?

Comment (ajo-odoo) [reply addons/mrp/report/mrp_report_bom_structure.py]: We have added this params as we need to hide column ( attachement ) in report when no attachement in bom line or sub bom line.

File: addons/mrp/report/mrp_report_bom_structure.py [hunk 8]
```diff
@@ -103,6 +103,8 @@ def _get_bom(self, bom_id=False, product_id=False, line_qty=False, line_id=False
             'level': level or 0,
             'operations': operations,
             'operations_cost': sum([op['total'] for op in operations]),
+            'attachments': self.env['mrp.document'].search(['|', '&',
+                ('res_model', '=', 'product.product'), ('res_id', '=', product.id), '&', ('res_model', '=', 'product.template'), ('res_id', '=', product.product_tmpl_id.id)]),
```
Comment (amoyaux) [inline addons/mrp/report/mrp_report_bom_structure.py]: @sle-odoo Is it worth to create an helper method that take the product.product as argument and return all the linked mrp.documents. Just to avoid this duplicated code?
